PLUGINS smooth-move and smooth-move-animations are a pair.

jon-heard opened the pull request:
Candle recently weighed in on the code-injection technique used in this plugin.  They had no qualms with it, seeing it as no different from rewriting the entire modified function in the plugin.
Also, I've split the plugin in two parts: smooth movement, and movement animations.  I've included both in this PR since they are so tied together.

jon-heard committed:
PLUGINS smooth-move and smooth-move-animations are a pair.

jon-heard commented:
I'm leaving `for (const dir of dirs) {` in smooth-move.  Snce it's a modification of an existing bipsi function, I don't want to change the logic unnecessarily.

seleb commented:
taking a look at this now, but quick reminder that candle is they/them

seleb committed:
eslint global

seleb committed:
disable eslint on for-of loop

seleb commented on src/smooth move animations.js:
```diff
@@ -0,0 +1,109 @@
+Builds on the smooth-move plugin to let you define unique walking and idle tiles in each direction
+for any event.  Useful if using bipsi for a more graphically involved game.
+
+NOTE - This plugin requires the smooth-move plugin in order to work.  It also needs to come AFTER
+smooth-move, so set it's "plugin-order" to something higher than what the smooth-move plugin's
+plugin-order is set to.
```
something to consider here: if this plugin requires `smooth move` and has some additional setup complexity bc of that, would it be better if it just included `smooth move` via es6 imports and the two were self-contained/mutually exclusive?

note that this isn't a blocker, just something to consider and there are pros/cons to each approach

jon-heard replied:
I'm not so familiar with this.  Are you suggesting to bundle `smooth-move` _into_ `smooth-move-animations`?  Or are you suggesting to replace the "plugin-order" step by having one imported plugin directly reference another imported plugin via es6 imports?

In terms of bundling: I could certainly do that.  My concern is that I think it might be more complex for the user to understand: one plugin does one thing, the other does the same thing, but also another thing.

seleb replied:
yeah, i'm suggesting they be bundled together

you're right that them including overlapping features may be confusing, but on the other hand this plugin is useless without the other and requires extra manual steps in order to set them up in a compatible way. if they were bundled together, the setup would be simplified

jon-heard replied:
Hmm.  I'm inclined to leave them separate since it's conceptually more straightforward, but I'll think further about this.

seleb committed:
switch disables to disable lines

jon-heard commented:
Oops, thanks for pointing out they/them.  I've adjusted that.

seleb commented on src/smooth move animations.js:
```diff
@@ -0,0 +1,109 @@
+Here is the full list of the tile fields smooth-move-animations can use:
+
+SMOOTH-MOVE-ANIMATIONS TILE FIELD KEYS:
+graphic-idle, graphic-move,
```
noticed a minor bug here: when `graphic-idle` is configured and the game is started, the player will use their regular `graphic` field until after they have moved once

jon-heard replied:
The problem is that the animation plugin does lazy setup for events.  This is triggered the first time an event moves, be that the avatar or a walking event (no distinction is made).  The pro is that only events that actually move are setup for movement.  The con is that they aren't setup until the first time they move.

The solution I landed on was to manually set each moving event's initial _regular_ `graphic` field to whatever tile the event should start with before first movement.  I absolutely forgot to mention this in the plugin's HOW TO, but I also wish there was a more user-friendly answer.  An alternative half-solution would be to have the animation plugin automatically setup the avatar's animation at game start.  It's only a half-solution since it doesn't fix the problem for non-avatar moving events.

Can you think of any better solutions?  Otherwise, I'm inclined to just add the manual-graphic-field "solution" to the HOW TO.

seleb replied:
thinking on this some more, the best solution here may actually be much simpler than expected: you could just remove `graphic-idle`. if i understand correctly, `graphic-idle` is redundant with the standard `graphic` field aside from this lazy setup issue, and it's only the move + directional idle fields that actually provide new functionality.

jon-heard replied:
This is a good idea.  'graphic' would need to replace both 'graphic-idle' and 'graphic-idle-down'.  The naming inconsistency is unfortunate.  However, given that the alternative is a field who's definition is meaningful only between game start and a moving event's first move, this seems reasonable.

Also, I am thinking that the documentation might not be explicit enough on one point.  You'd generally either use 'graphic-idle' and 'graphic-move' OR you'd use 'graphic-idle-down', 'graphic-idle-up', ... 'graphic-move-left', 'graphic-move-right'.  The directions are the most detailed setup and the idle/move is a simper alternative.  I'll update the documentation to make that more explicit.

seleb reviewed (changes requested):
minor issue with smooth move animations, but otherwise lgtm

jon-heard committed:
- Changed "smooth move animations.js" to "move animations.js"
- Decoupled "move animations.js" from "smooth move.js" (with optional coupling by user).
- Rewrote documentation.

jon-heard commented:
I've decoupled "smooth move animations" from "smooth move" and renamed it "move animations".  It still has optional extra functionality if combined with "smooth move" but now works on its own as well.

seleb committed:
Merge branch 'main' into newPlugin_smoothMove

seleb committed:
disable eslint on line

seleb commented:
:tada: This PR is included in version 4.12.0 :tada:

The release is available on [GitHub release](https://github.com/seleb/bipsi-hacks/releases/tag/v4.12.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
